pkg/storage/stores/tsdb/cacheable_index.go

```diff
@@ -0,0 +1,436 @@
+			Namespace: "loki",
+			Name:      "tsdb_index_cache_put_errors_total",
+			Help:      "The number of puts for the tsdb index cache that failed.",
+		}),
+	}
+
+	if reg != nil {
+		reg.MustRegister(m.cacheDecodeErrs)
+		reg.MustRegister(m.cacheHits)
+		reg.MustRegister(m.cacheMisses)
+		reg.MustRegister(m.cacheGets)
+		reg.MustRegister(m.cacheGetErrors)
+		reg.MustRegister(m.cachePuts)
+		reg.MustRegister(m.cachePutErrors)
+		reg.MustRegister(m.cacheEncodeErrs)
+	}
+
+	return m
+}
+
+// NewCacheableIndex instantiate a new cacheable index given an index and a cache implementation.
+//
+// It assumes the cache is ready to be used.
+func NewCacheableIndex(l log.Logger, index *TSDBIndex, cache cache.Cache, m CacheableIndexMetrics) CacheableIndex {
+	return CacheableIndex{
+		Index:  index,
+		Cache:  cache,
+		m:      m,
+		logger: l,
+	}
+}
+
+// Stop does everything necessary to stop this index safely.
+func (i *CacheableIndex) Stop() {
+	i.Cache.Stop()
+}
+
+// stringifiedMatchers build and return a promQL string based on the union
+// of all the given matchers.
+func stringifiedMatchers(matchers []*labels.Matcher) string {
+	b := strings.Builder{}
+
+	b.WriteByte('{')
+	for i, m := range matchers {
+		if i > 0 {
+			b.WriteByte(',')
+			b.WriteByte(' ')
+		}
+
+		b.WriteString(m.String())
+	}
+	b.WriteByte('}')
+
+	return b.String()
+}
+
+// buildCacheKey constructs the key used to retrieve and store values from/into the cache.
+//
+// It joins all parameters with the collon operator to avoid collisions.
+func (i *CacheableIndex) buildCacheKey(userID string, from, through model.Time, shard *index.ShardAnnotation, matchers ...*labels.Matcher) string {
+	stringifiedShard := fmt.Sprint(shard)
+	keyMembers := []string{
+		userID, from.String(), through.String(), stringifiedShard,
+	}
+	if len(matchers) != 0 {
+		keyMembers = append(keyMembers, stringifiedMatchers(matchers))
+	}
+
+	return strings.Join(keyMembers, sep)
+}
+
+// GetChunkRefs is a cached implementation of GetChunkRefs.
+//
+// It uses as a key for the cache all parameters (userID, from, through, shard, matchers) separated by a colon (:).
+func (i *CacheableIndex) GetChunkRefs(ctx context.Context, userID string, from, through model.Time, res []ChunkRef, shard *index.ShardAnnotation, matchers ...*labels.Matcher) ([]ChunkRef, error) {
+	fallbackFn := func(ctx context.Context, userID string, from, through model.Time, shard *index.ShardAnnotation, matcher ...*labels.Matcher) ([]ChunkRef, []byte, error) {
+		chunks, err := i.Index.GetChunkRefs(ctx, userID, from, through, res, shard, matcher...)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "call to GetChunkRefs")
+		}
+
+		buf, err := proto.Marshal(&ChunkRefs{chunks})
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "Series encoding")
+		}
+
+		return chunks, buf, nil
+	}
+
+	key := i.buildCacheKey(userID, from, through, shard, matchers...)
+	keys := []string{key}
+	_, response, misses := i.cacheFetch(ctx, keys)
```

Comment:
I'm trying to understand what are the chance we hit this cache between two index query.

For instance, from 10 to  20, then we do from 0 to 20 ? 

e.g what are the chance that from and through are exactly equal ? Sounds like never ? Am I missing something ?

I'd like to also compare this with Mimir implementation ?  Sounds like they are caching at a lower level to probably increase the chance of hitting the cache between multiple queries.

see https://github.com/grafana/mimir/blob/main/pkg/storegateway/indexcache/cache.go#L42-L76

@pracucci @pstibrany Anything we should know about your index cache ? What works well ? what doesn't ?

Reply:
the chances are fairly low. The idea was to first implement this initial version that rarely hits the cache and then implement the optimization of splitting the timestamp into smaller fixed steps that can be reused by other queries.

The optimization in detail would be:
If the step configured for caching is 10 and you sent a query from 20 to 40, later if you sent a query from 10 to 30, the data between 20~30 will be cached so you'll only have to fetch data from 10 to 20.

Reply:
I haven't reviewed this PR (missing too much context) but Mimir index cache is "high level" from my perspective. We do cache the actual lookup result and not the index sub-object portions (like we do for the chunks cache).

The index cache is very effective in Mimir. For example, looking up the cache hit ratio over the last 24h across all Mimir clusters running at Grafana Labs, we get:

```
sum by(item_type) (rate(thanos_store_index_cache_hits_total[$__range])) /
sum by(item_type) (rate(thanos_store_index_cache_requests_total[$__range]))
* 100
```

![Screenshot 2022-05-03 at 12 52 26](https://user-images.githubusercontent.com/1701904/166441119-8652b128-7902-4c15-9ad1-93b6b34b9fa1.png)

Reply:
The cache hit rate for Mimir is good because they are caching data at block level which we can't do at least for live streams since our index shards are just 15m long so our index files for live data change much often.

I think we should check the feasibility of the proposed follow-up to improve the hit rate i.e smaller steps for lookup and merging the results. This approach works well for the results cache because we do not set any expiry on the cached results and let the underlying cache drop the items if it runs out of memory. I am wondering if this approach would work well for index cache since the expiry is just 5m. If we can't do it or it doesn't work out well then the cache would not be useful at all. Let me see if I can put some thoughts in a doc on what we can do about it.